### DISCUSSION TITLE
Keycloak registration/signup page

### DISCUSSION THREAD

**Discussion by syam5492009** [nextauthjs/next-auth#6368] [Help]
### Question 💬

Thanks for the awesome library. Implemented the keycloak authentication with next-auth. In one condition, i need to redirect the user to keycloak registration page instead of signIn page. How can i acheive it

### How to reproduce ☕️

openid-connect/registrations - custom link used . But no use. please help me on this

### Contributing 🙌🏽

Yes, I am willing to help answer this question in a PR

**Comment by balazsorban44** [comment]
If this is the right spec, you just set the `prompt: "create"` parameter for the authorization endpoint:

https://openid.net/specs/openid-connect-prompt-create-1_0.html#name-prompt-parameter
https://next-auth.js.org/configuration/providers/oauth#override-default-options
https://next-auth.js.org/getting-started/client#additional-parameters

**Reply by syam5492009** [reply]
I did used like this .    **signIn("keycloak", null, { prompt: "create" });**
But it is redirecting me to Login Page only.

**Reply by V-Mann-Nick** [reply]
Yeah this unfortunately does not work.

Keylcoak serves authorization endpoint on `{origin}/auth/realms/{realm}/protocol/openid-connect/auth` while registration is on `{origin}/auth/realms/{realm}/protocol/openid-connect/registrations`.

`?prompt=create` is also not respected by Keycloak as @syam5492009 noted.

I could of course create the registration link manually but then I'm unsure on how to check for state and nonce on callback. The registration works this way but the redirect to `/api/auth/callback/keycloak` leads to an error.

I tried using a custom `authorization.request` which typescript told me about, but that apparently is a dud as you [have noted](https://github.com/nextauthjs/next-auth/issues/4842) @balazsorban44.

Only way to do registrations properly is going to the login of Keycloak and clicking "Register" there, but that does not fit our requirements as we need a Keycloak register button in our UI.

**Comment by V-Mann-Nick** [comment]
So I have found a very hack workaround, I'm not very happy with at all but I guess it will have to do for now.

First I created a function

```ts
const registrationEndpoint = `${process.env.NEXT_PUBLIC_KEYCLOAK_URL}/realms/${process.env.NEXT_PUBLIC_KEYCLOAK_REALM}/protocol/openid-connect/registrations`

export const createRegistrationUrl = (redirectUri: string) => {
  const params = new URLSearchParams()
  params.set('client_id', process.env.NEXT_PUBLIC_KEYCLOAK_CLIENT_ID)
  params.set('redirect_uri', redirectUri)
  params.set('scope', 'openid profile email')
  params.set('response_type', 'code')
  params.set('response_mode', 'fragment')
  return `${registrationEndpoint}?${params.toString()}`
}
```

This is mostly taken from [keycloak-js](https://github.com/keycloak/keycloak/blob/main/adapters/oidc/js/src/keycloak.js#L390) although it's vastly simplified and I'm sure there's some security aspect of it I'm not understanding and completely neglecting.

I then use this function like so:

```ts
createRegistrationUrl(`${window.location.origin}/callback?type=post_register?redirect_uri=/`)
```

Note the following:
- `window.location.origin` can't be used on the server and might cause a hydration mismatch. I used an env variable `NEXT_PUBLIC_HOST=http://localhost:300` to avoid that
- `?type=post_register&redirect_uri=/` will be used by the `/callback` page

This Page goes ruffly like this:

```ts
export default function Callback() {
  const router = useRouter()
  useEffect(() => {
    if (!router.isReady) return
    const type = assertString(router.query['type'])
    const redirectUri = assertString(router.query['redirect_uri'])
    if (!type) {
      router.replace(redirectUri ?? '/')
      return
    }
    switch (type) {
      case 'post_register': {
        if (!redirectUri) {
          router.replace('/')
          return
        }
        signIn('keycloak', { callbackUrl: redirectUri })
        return
      }
    }
  }, [router])
  return <span>Loading...</span>
  )
}
```

This signes in the user on the next-auth side and as the user is already logged in on the keycloak side there's no need to login. Next-auth will then direct the user back to the page they originally came from with the `callbackUrl` option.

I guess this is acceptable although I would love to see a way to accomplish this in next-auth.

**Comment by SultanNasyrovDeveloper** [comment]
+1

**Comment by liamlows** [comment]
As far as I can tell, the easiest path forward would be for Keycloak to support the OIDC `prompt=create`, there is currently an issue open here: https://github.com/keycloak/keycloak/issues/10701

Hopefully Keycloak will notice it and add it (seems to be a relatively simple add) although there has been no response from devs thus far in 2 years since it was opened.

**Reply by kirtimaan-askin** [reply]
Seems that https://github.com/keycloak/keycloak/issues/10701 was addressed and merged into keycloak 26.x version. 

Unless it's already somehow addressed in next-auth, may I take this up and implement as contribution to this awesome package?

**Reply by Tmauc** [reply]
Hi, do you have any news about the implementation of prompt=create in nextauth?

**Reply by DerIch27** [reply]
you can use `await signIn("keycloak", null, {prompt: "create"}`
